core/src/main/scala/kafka/server/metadata/RaftMetadataCache.scala

```diff
@@ -0,0 +1,390 @@
+        case None =>
+          Node.noNode
+      }
+    }
+  }
+
+  override def getPartitionReplicaEndpoints(tp: TopicPartition, listenerName: ListenerName): Map[Int, Node] = {
+    val image = _currentImage
+    image.partitions.topicPartition(tp.topic(), tp.partition()).map { partition =>
+      partition.replicas.asScala.map(replicaId => replicaId.intValue() -> {
+        image.aliveBroker(replicaId) match {
+          case Some(broker) =>
+            broker.endpoints.getOrElse(listenerName.value(), Node.noNode())
+          case None =>
+            Node.noNode()
+        }}).toMap
+        .filter(pair => pair match {
+          case (_, node) => !node.isEmpty
+        })
+    }.getOrElse(Map.empty[Int, Node])
+  }
+
+  override def getControllerId: Option[Int] = {
+    _currentImage.controllerId
+  }
+
+  override def getClusterMetadata(clusterId: String, listenerName: ListenerName): Cluster = {
+    val image = _currentImage
+    val nodes = new util.HashMap[Integer, Node]
+    image.brokers.aliveBrokers().foreach { node =>
+      if (!node.fenced) {
+        node.endpoints.get(listenerName.value()).foreach { nodes.put(node.id, _) }
+      }
+    }
+
+    def node(id: Integer): Node = {
+      Option(nodes.get(id)).getOrElse(new Node(id, "", -1))
+    }
+
+    val partitionInfos = new util.ArrayList[PartitionInfo]
+    val internalTopics = new util.HashSet[String]
+
+    image.partitions.allPartitions().foreach { partition =>
+      partitionInfos.add(new PartitionInfo(partition.topicName,
+        partition.partitionIndex, node(partition.leaderId),
+        partition.replicas.asScala.map(node).toArray,
+        partition.isr.asScala.map(node).toArray,
+        partition.offlineReplicas.asScala.map(node).toArray))
+      if (Topic.isInternal(partition.topicName)) {
+        internalTopics.add(partition.topicName)
+      }
+    }
+
+    new Cluster(clusterId, nodes.values(),
+      partitionInfos, Collections.emptySet[String], internalTopics,
+      node(Integer.valueOf(image.controllerId.getOrElse(-1))))
+  }
+
+  def stateChangeTraceEnabled(): Boolean = {
+    stateChangeLogger.isTraceEnabled
+  }
+
+  def logStateChangeTrace(str: String): Unit = {
+    stateChangeLogger.trace(str)
+  }
+
+  // This method returns the deleted TopicPartitions received from UpdateMetadataRequest
+  override def updateMetadata(correlationId: Int, request: UpdateMetadataRequest): Seq[TopicPartition] = {
+    inLock(lock) {
+      val image = _currentImage
+      val builder = MetadataImageBuilder(brokerId, logger.underlying, image)
+
+      builder.controllerId(if (request.controllerId() < 0) None else Some(request.controllerId()))
+
+      // Compare the new brokers with the existing ones.
+      def toMetadataBroker(broker: UpdateMetadataBroker): MetadataBroker = {
+        val endpoints = broker.endpoints().asScala.map { endpoint =>
+          endpoint.listener -> new Node(broker.id(), endpoint.host(), endpoint.port())
```

Comment:
Note that here we are, somewhat incorrectly, excluding the broker's rack from the Node object. This was done to keep the behavior compatible with the existing MetadataCache (now ZkMetadataCache). Once we have this merged and 2.8 is released, we can work on determining if this was done for a good reason, or if it's a bug in the existing MetadataCache logic. Since this class isn't used for any production code yet, we can leave this as-is.